Client/Application/BackgroundFirefoxSuggestIngestUtility.swift

```diff
@@ -0,0 +1,86 @@
+import Foundation
+import Shared
+import Storage
+
+/// A background utility that downloads and stores new Firefox Suggest
+/// suggestions when the device is online and connected to power.
+class BackgroundFirefoxSuggestIngestUtility: BackgroundUtilityProtocol {
+    static let taskIdentifier = "org.mozilla.ios.firefox.suggest.ingest"
+
+    let firefoxSuggest: RustFirefoxSuggest
+    let logger: Logger
+
+    init(firefoxSuggest: RustFirefoxSuggest, logger: Logger = DefaultLogger.shared) {
+        self.firefoxSuggest = firefoxSuggest
+        self.logger = logger
+
+        setUp()
+    }
+
+    /// Schedules the ingestion task to run when the app is backgrounded.
+    func scheduleTaskOnAppBackground() {
+        let request = createTaskRequest()
+        do {
+            try BGTaskScheduler.shared.submit(request)
+        } catch {
+            logger.log("Failed to schedule background ingestion: \(error.localizedDescription)",
+                       level: .warning,
+                       category: .storage)
+        }
+    }
+
+    /// Create a request to schedule the background ingestion task.
+    private func createTaskRequest() -> BGProcessingTaskRequest {
+        let request = BGProcessingTaskRequest(identifier: Self.taskIdentifier)
+        request.requiresNetworkConnectivity = true
+        request.requiresExternalPower = true
+        return request
+    }
+
+    /// Downloads and stores new Firefox Suggest suggestions. Returns `true` on
+    /// success or `false` on failure.
+    private func ingest() async -> Bool {
+        logger.log("Ingesting new suggestions",
+                   level: .debug,
+                   category: .storage)
+        do {
+            try await firefoxSuggest.ingest()
+            logger.log("Successfully ingested new suggestions",
+                       level: .debug,
+                       category: .storage)
+            return true
+        } catch {
+            logger.log("Failed to ingest new suggestions: \(error.localizedDescription)",
+                       level: .warning,
+                       category: .storage)
+            return false
+        }
+    }
+
+    /// Registers a launch handler for the background ingestion task.
+    private func setUp() {
+        BGTaskScheduler.shared.register(forTaskWithIdentifier: Self.taskIdentifier, using: nil) { task in
+            Task {
+                let success = await self.ingest()
+                if !success {
+                    // If ingestion failed, schedule a follow-up task to retry.
+                    let request = self.createTaskRequest()
+                    do {
+                        try BGTaskScheduler.shared.submit(request)
+                    } catch {
+                        self.logger.log("Failed to schedule retry ingestion: \(error.localizedDescription)",
+                                        level: .warning,
+                                        category: .storage)
+                    }
```

Comment:
Hehe, I like it! I changed this method a bit, to also submit the request to `BGTaskScheduler`. Originally, I had it take an optional `minimumStartDelay` parameter, but got rid of that, so creating the request separately doesn't make a whole lot of sense anymore.

I did keep the `do`/`catch` and logging in the callers, because I have a slight preference for logging directly, instead of passing (parts of) log messages around. Doing it that way feels easier to match up logs to code by grepping for a known log string—but I don't feel too strongly about it, and also happy to change it! 😄